webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/facts/nodes/NodeFactRepository.scala

```diff
@@ -631,54 +631,60 @@ class CoreNodeFactRepository(
-                      // we also keep status
-                      case (Left(x), Some(y))       =>
-                        x
-                          .modify(_.rudderSettings.security)
-                          .setTo(y.rudderSettings.security)
-                          .modify(_.rudderSettings.status)
-                          .setTo(y.rudderSettings.status)
-                          .succeed
-                      // second case: we are updating tenants
-                      case (Right((_, t)), Some(x)) => NodeFact.fromMinimal(x.modify(_.rudderSettings.security).setTo(t)).succeed
-                      case (Right((id, _)), None)   =>
-                        Inconsistency(s"Error: can not change tenant of missing node with Id '${id.value}'").fail
-                    }
-        es       <- tenantService.manageUpdate(n, nodeFact, cc) { updated =>
-                      for {
-                        // first we persist on cold storage, which is more likely to fail. Plus, for history reason, some
-                        // mapping are not exactly isomorphic, and some normalization can happen - for ex, for missing machine.
-                        s <- storage.save(updated)
-                        // then, we get the actual thing that was saved from the save event
-                        up = s match {
-                               case StorageChangeEventSave.Created(node, attrs)             =>
-                                 node
-                               case StorageChangeEventSave.Updated(oldNode, newNode, attrs) =>
-                                 newNode
-                               case StorageChangeEventSave.Noop(nodeId, attrs)              =>
-                                 updated
-                             }
-                        // then, we save that as the core node fact reference
-                        e <- nodeFact.rudderSettings.status match {
-                               case RemovedInventory  => // this case is ignored, we don't delete node based on status value
-                                 StorageChangeEventSave.Noop(up.id, attrs).succeed
-                               case PendingInventory  =>
-                                 saveOn(pendingNodes, up.toCore)
-                               case AcceptedInventory =>
-                                 saveOn(acceptedNodes, up.toCore)
-                             }
-                        // finally, merge both events and return the final change event and run call backs on it
-                        es = s.updateWith(e).toChangeEvent(up.id, up.rudderSettings.status, cc)
-                        _ <- NodeLoggerPure.Details.Write.trace(s"Saved node event: ${es.event.debugDiff}")
-                        _ <- runCallbacks(es)
-                      } yield es
-                    }
+        _            <- lock.withLock
+        core         <- get(node.fold(_.id, _._1))(cc.toQuery)
+        pair         <- (node, core) match {
```

Comment:
Instead of pair, can you do 

(nodeFact, a) <- pair

in fact instead of a, rename to something meaningful (like attributes or attr)

Reply:
If scala compiler does not prevent you to do so, sometimes it does not like this ...

Reply:
it was necessary because zio doesn',t have a filtering method